src/boss/lambdastore/store.go

```diff
@@ -188,34 +182,42 @@ func (s *LambdaStore) unregisterTriggers(functionName string) {
 // assumes the caller holds the function lock
 func (s *LambdaStore) addToRegistry(name string, body io.Reader) error {
 	tarPath := filepath.Join(s.StorePath, name+".tar.gz")
+	tmpPath := tarPath + ".tmp"
 
-	tarFile, err := os.Create(tarPath)
+	tarFile, err := os.Create(tmpPath)
 	if err != nil {
-		return fmt.Errorf("failed to create lambda tarball: %w", err)
+		return fmt.Errorf("failed to create temp tarball: %w", err)
 	}
-	defer tarFile.Close()
 
-	// TODO: atomically write tarball by writing to a temp file and renaming
 	if _, err := io.Copy(tarFile, body); err != nil {
-		return fmt.Errorf("failed to write tarball: %w", err)
+		tarFile.Close()
+		_ = os.Remove(tmpPath)
```

Comment:
Rather than have os.Remove each time, we can have a defer with an anonymous function that always tries to do Remove when addToRegistry returns.